azurestack/data_source_public_ip.go

```diff
@@ -0,0 +1,84 @@
+package azurestack
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurestack/azurestack/utils"
+)
+
+func dataSourceArmPublicIP() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceArmPublicIPRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"resource_group_name": resourceGroupNameForDataSourceSchema(),
+
+			"domain_name_label": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"idle_timeout_in_minutes": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"fqdn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"ip_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func dataSourceArmPublicIPRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).publicIPClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	resp, err := client.Get(ctx, resGroup, name, "")
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Error: Public IP %q (Resource Group %q) was not found", name, resGroup)
+		}
+		return fmt.Errorf("Error making Read request on Azure public ip %s: %s", name, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	if resp.PublicIPAddressPropertiesFormat.DNSSettings != nil {
```

Comment:
This could be vastly easier to read (as well have additional nil checking):

```go
if props := resp.PublicIPAddressPropertiesFormat; props != nil {
		if dnsSettings := props.DNSSettings; dnsSettings != nil {
			if v := dnsSettings.Fqdn; v != nil && *v != "" {
				d.Set("fqdn", v)
			}

			if v := dnsSettings.DomainNameLabel; v != nil && *v != "" {
				d.Set("domain_name_label", v)
			}
		}
	}
```